File: backend/internal/idp/attribute_configuration_test.go
/*
 * Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package idp

import (
	"context"

	"github.com/stretchr/testify/mock"

	"github.com/thunder-id/thunderid/internal/entitytype"
	"github.com/thunder-id/thunderid/internal/system/error/serviceerror"
	"github.com/thunder-id/thunderid/internal/system/i18n/core"
)

// singleProfileMapping builds an attribute configuration that resolves to userType with a single
// user-type-attributes entry carrying the given claim mappings.
func singleProfileMapping(userType string, mappings []AttributeMapping) *AttributeConfiguration {
	return &AttributeConfiguration{
		UserTypeResolution:        &UserTypeResolution{Default: userType},
		UserTypeAttributeMappings: []UserTypeAttributeMapping{{UserType: userType, Attributes: mappings}},
	}
}

func (s *IDPServiceTestSuite) TestValidateAttributeConfiguration_NilMapping_OK() {
	svcErr := s.idpService.validateAttributeConfiguration(context.Background(), &IDPDTO{})
	s.Nil(svcErr)
}

func (s *IDPServiceTestSuite) TestValidateAttributeConfiguration_Valid() {
	s.mockET.On("GetAttributes", mock.Anything, entitytype.TypeCategoryUser, "person", false, true, false).
		Return([]entitytype.AttributeInfo{{Attribute: "firstName"}, {Attribute: "email"}},
			(*serviceerror.ServiceError)(nil))

	idp := &IDPDTO{AttributeConfiguration: singleProfileMapping("person", []AttributeMapping{
		{ExternalAttribute: "given_name", LocalAttribute: "firstName"},
		{ExternalAttribute: "address.email", LocalAttribute: "email"},
	})}

	s.Nil(s.idpService.validateAttributeConfiguration(context.Background(), idp))
}

func (s *IDPServiceTestSuite) TestValidateAttributeConfiguration_EmptyUserType() {
	idp := &IDPDTO{AttributeConfiguration: &AttributeConfiguration{
		UserTypeAttributeMappings: []UserTypeAttributeMapping{{
			Attributes: []AttributeMapping{{ExternalAttribute: "given_name", LocalAttribute: "firstName"}},
		}},
	}}
	svcErr := s.idpService.validateAttributeConfiguration(context.Background(), idp)
	s.NotNil(svcErr)
	s.Equal(ErrorInvalidAttributeConfiguration.Code, svcErr.Code)
}

func (s *IDPServiceTestSuite) TestValidateAttributeConfiguration_EmptyMappings() {
	s.mockET.On("GetAttributes", mock.Anything, entitytype.TypeCategoryUser, "person", false, true, false).
		Return([]entitytype.AttributeInfo{{Attribute: "firstName"}}, (*serviceerror.ServiceError)(nil))
	idp := &IDPDTO{AttributeConfiguration: singleProfileMapping("person", nil)}
	svcErr := s.idpService.validateAttributeConfiguration(context.Background(), idp)
	s.Nil(svcErr)
}

func (s *IDPServiceTestSuite) TestValidateAttributeConfiguration_OneSourceToMultipleTargets() {
	s.mockET.On("GetAttributes", mock.Anything, entitytype.TypeCategoryUser, "person", false, true, false).
		Return([]entitytype.AttributeInfo{{Attribute: "email"}, {Attribute: "contactEmail"}},
			(*serviceerror.ServiceError)(nil))

	idp := &IDPDTO{AttributeConfiguration: singleProfileMapping("person", []AttributeMapping{
		{ExternalAttribute: "email", LocalAttribute: "email"},
		{ExternalAttribute: "email", LocalAttribute: "contactEmail"},
	})}

	s.Nil(s.idpService.validateAttributeConfiguration(context.Background(), idp))
}

func (s *IDPServiceTestSuite) TestValidateAttributeConfiguration_DuplicateTarget() {
	idp := &IDPDTO{AttributeConfiguration: singleProfileMapping("person", []AttributeMapping{
		{ExternalAttribute: "given_name", LocalAttribute: "firstName"},
		{ExternalAttribute: "first_name", LocalAttribute: "firstName"},
	})}
	svcErr := s.idpService.validateAttributeConfiguration(context.Background(), idp)
	s.NotNil(svcErr)
	s.Equal(ErrorInvalidAttributeConfiguration.Code, svcErr.Code)
	s.Contains(svcErr.ErrorDescription.DefaultValue, "more than once")
}

func (s *IDPServiceTestSuite) TestValidateAttributeConfiguration_DuplicateUserType() {
	s.mockET.On("GetAttributes", mock.Anything, entitytype.TypeCategoryUser, "person", false, true, false).
		Return([]entitytype.AttributeInfo{{Attribute: "firstName"}}, (*serviceerror.ServiceError)(nil))
	idp := &IDPDTO{AttributeConfiguration: &AttributeConfiguration{
		UserTypeResolution: &UserTypeResolution{Default: "person"},
		UserTypeAttributeMappings: []UserTypeAttributeMapping{
			{
				UserType:   "person",
				Attributes: []AttributeMapping{{ExternalAttribute: "given_name", LocalAttribute: "firstName"}},
			},
			{
				UserType:   "person",
				Attributes: []AttributeMapping{{ExternalAttribute: "family_name", LocalAttribute: "lastName"}},
			},
		},
	}}
	svcErr := s.idpService.validateAttributeConfiguration(context.Background(), idp)
	s.NotNil(svcErr)
	s.Equal(ErrorInvalidAttributeConfiguration.Code, svcErr.Code)
	s.Contains(svcErr.ErrorDescription.DefaultValue, "configured more than once")
}

func (s *IDPServiceTestSuite) TestValidateAttributeConfiguration_TargetNotInSchema() {
	s.mockET.On("GetAttributes", mock.Anything, entitytype.TypeCategoryUser, "person", false, true, false).
		Return([]entitytype.AttributeInfo{{Attribute: "email"}}, (*serviceerror.ServiceError)(nil))

	idp := &IDPDTO{AttributeConfiguration: singleProfileMapping("person", []AttributeMapping{
		{ExternalAttribute: "given_name", LocalAttribute: "firstName"},
	})}
	svcErr := s.idpService.validateAttributeConfiguration(context.Background(), idp)
	s.NotNil(svcErr)
	s.Equal(ErrorInvalidAttributeConfiguration.Code, svcErr.Code)
	s.Contains(svcErr.ErrorDescription.DefaultValue, "not an attribute")
}

func (s *IDPServiceTestSuite) TestValidateAttributeConfiguration_UnknownUserType() {
	s.mockET.On("GetAttributes", mock.Anything, entitytype.TypeCategoryUser, "ghost", false, true, false).
		Return([]entitytype.AttributeInfo(nil), &serviceerror.ServiceError{
			Type: serviceerror.ClientErrorType, Code: "ETS-1004",
			ErrorDescription: core.I18nMessage{DefaultValue: "entity type not found"},
		})

	idp := &IDPDTO{AttributeConfiguration: singleProfileMapping("ghost", []AttributeMapping{
		{ExternalAttribute: "given_name", LocalAttribute: "firstName"},
	})}
	svcErr := s.idpService.validateAttributeConfiguration(context.Background(), idp)
	s.NotNil(svcErr)
	s.Equal(ErrorInvalidAttributeConfiguration.Code, svcErr.Code)
}

File: frontend/packages/configure-connections/src/utils/__tests__/attributeConfiguration.test.ts
/**
 * Copyright (c) 2025, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import type {PropertyDefinition} from '@thunderid/configure-user-types';
import {describe, expect, it} from 'vitest';
import type {AttributeConfiguration} from '../../models/connection';
import {
  type AttributeMappingFormState,
  flattenUserTypeAttributes,
  fromAttributeConfiguration,
  toAttributeConfiguration,
} from '../attributeConfiguration';

const emptyState: AttributeMappingFormState = {
  defaultUserType: '',
  resolveDynamic: false,
  externalAttribute: '',
  valueMapping: [],
  groups: [],
  linking: [],
};

describe('toAttributeConfiguration', () => {
  it('returns undefined when the whole configuration is empty', () => {
    expect(toAttributeConfiguration(emptyState)).toBeUndefined();
  });

  it('builds a static default-only config', () => {
    expect(toAttributeConfiguration({...emptyState, defaultUserType: ' Person '})).toEqual({
      userTypeResolution: {default: 'Person'},
    });
  });

  it('builds per-user-type mappings, dropping incomplete rows and empty groups, trimming values', () => {
    const cfg = toAttributeConfiguration({
      ...emptyState,
      defaultUserType: 'Person',
      groups: [
        {
          userType: ' Person ',
          rows: [
            {externalAttribute: ' given_name ', localAttribute: ' firstName '},
            {externalAttribute: 'email', localAttribute: ''},
          ],
        },
        {userType: 'Employee', rows: [{externalAttribute: '', localAttribute: ''}]},
      ],
    });
    expect(cfg).toEqual({
      userTypeResolution: {default: 'Person'},
      userTypeAttributeMappings: [
        {userType: 'Person', attributes: [{externalAttribute: 'given_name', localAttribute: 'firstName'}]},
      ],
    });
  });

  it('includes the external attribute and drops incomplete value-mapping entries', () => {
    const cfg = toAttributeConfiguration({
      ...emptyState,
      defaultUserType: 'Person',
      resolveDynamic: true,
      externalAttribute: ' user_type ',
      valueMapping: [
        {value: ' staff ', userType: ' Employee '},
        {value: 'incomplete', userType: ''},
      ],
    });
    expect(cfg).toEqual({
      userTypeResolution: {default: 'Person', externalAttribute: 'user_type', valueMapping: {staff: 'Employee'}},
    });
  });

  it('includes the external attribute alone when no value mappings are configured', () => {
    const cfg = toAttributeConfiguration({
      ...emptyState,
      defaultUserType: 'Person',
      resolveDynamic: true,
      externalAttribute: 'user_type',
      valueMapping: [],
    });
    expect(cfg).toEqual({userTypeResolution: {default: 'Person', externalAttribute: 'user_type'}});
  });

  it('omits dynamic fields when the toggle is off even if value mappings linger in state', () => {
    const cfg = toAttributeConfiguration({
      ...emptyState,
      defaultUserType: 'Person',
      resolveDynamic: false,
      externalAttribute: 'user_type',
      valueMapping: [{value: 'staff', userType: 'Employee'}],
    });
    expect(cfg).toEqual({userTypeResolution: {default: 'Person'}});
  });

  it('includes account linking only for non-empty trimmed attributes', () => {
    const cfg = toAttributeConfiguration({...emptyState, defaultUserType: 'Person', linking: [' email ', '', '  ']});
    expect(cfg).toEqual({userTypeResolution: {default: 'Person'}, accountLinking: {attributes: ['email']}});
  });
});

describe('fromAttributeConfiguration', () => {
  it('returns empty state for undefined config', () => {
    expect(fromAttributeConfiguration(undefined)).toEqual(emptyState);
  });

  it('hydrates all three sections and infers the dynamic toggle', () => {
    const cfg: AttributeConfiguration = {
      userTypeResolution: {
        default: 'Person',
        externalAttribute: 'user_type',
        valueMapping: {staff: 'Employee'},
      },
      userTypeAttributeMappings: [
        {userType: 'Person', attributes: [{externalAttribute: 'given_name', localAttribute: 'firstName'}]},
      ],
      accountLinking: {attributes: ['email']},
    };
    expect(fromAttributeConfiguration(cfg)).toEqual({
      defaultUserType: 'Person',
      resolveDynamic: true,
      externalAttribute: 'user_type',
      valueMapping: [{value: 'staff', userType: 'Employee'}],
      groups: [{userType: 'Person', rows: [{externalAttribute: 'given_name', localAttribute: 'firstName'}]}],
      linking: ['email'],
    });
  });

  it('round-trips a rich state with toAttributeConfiguration', () => {
    const state: AttributeMappingFormState = {
      defaultUserType: 'Person',
      resolveDynamic: true,
      externalAttribute: 'user_type',
      valueMapping: [{value: 'staff', userType: 'Employee'}],
      groups: [
        {userType: 'Person', rows: [{externalAttribute: 'email', localAttribute: 'email'}]},
        {userType: 'Employee', rows: [{externalAttribute: 'emp_id', localAttribute: 'employeeNumber'}]},
      ],
      linking: ['email'],
    };
    expect(fromAttributeConfiguration(toAttributeConfiguration(state))).toEqual(state);
  });
});

describe('flattenUserTypeAttributes', () => {
  it('skips credential + array attributes and recurses objects with dot-notation', () => {
    const schema = {
      firstName: {type: 'string'},
      password: {type: 'string', credential: true},
      roles: {type: 'array'},
      address: {type: 'object', properties: {email: {type: 'string'}, zip: {type: 'number'}}},
    } as unknown as Record<string, PropertyDefinition>;

    expect(flattenUserTypeAttributes(schema).sort()).toEqual(['address.email', 'address.zip', 'firstName']);
  });

  it('returns an empty list for undefined schema', () => {
    expect(flattenUserTypeAttributes(undefined)).toEqual([]);
  });
});
